### PR TITLE
[MRG]: allow n_dimensional search lights

### PR DESCRIPTION
**Pull request by kingjr**
This allows using n-dimensional arrays with the search light, rather than being constrained to 3D array.

This is the first step to before being able to have a rolling window with more than 1 time point.

### CODE REVIEW

**Comment by kingjr** [comment]
ready for review

File: mne/decoding/search_light.py [hunk 1]
```diff
@@ -61,9 +63,11 @@ def fit(self, X, y):
 
         Parameters
         ----------
-        X : array, shape (n_samples, n_features, n_estimators)
+        X : array, shape (n_samples, nd_features, n_estimators)
             The training input samples. For each data slice, a clone estimator
-            is fitted independently.
+            is fitted independently.The feature dimension can be
```
Comment (agramfort) [inline mne/decoding/search_light.py]: space after .

File: mne/decoding/search_light.py [hunk 2]
```diff
@@ -76,6 +80,7 @@ def fit(self, X, y):
         self.estimators_ = list()
         # For fitting, the parallelization is across estimators.
         parallel, p_func, n_jobs = parallel_func(_sl_fit, self.n_jobs)
+        n_jobs = np.min([n_jobs, X.shape[-1]])
```
Comment (agramfort) [inline mne/decoding/search_light.py]: np.min([n_jobs, X.shape[-1]])

->

min(n_jobs, X.shape[-1])

using builtin

File: mne/decoding/search_light.py [hunk 3]
```diff
@@ -88,10 +93,11 @@ def _transform(self, X, method):
         method = _check_method(self.base_estimator, method)
         if X.shape[-1] != len(self.estimators_):
             raise ValueError('The number of estimators does not match '
-                             'X.shape[2]')
+                             'X.shape[-1]')
         # For predictions/transforms the parallelization is across the data and
         # not across the estimators to avoid memory load.
         parallel, p_func, n_jobs = parallel_func(_sl_transform, self.n_jobs)
+        n_jobs = np.min([n_jobs, X.shape[-1]])
```
Comment (agramfort) [inline mne/decoding/search_light.py]: idem

File: mne/decoding/search_light.py [hunk 4]
```diff
@@ -207,10 +225,11 @@ def score(self, X, y):
         self._check_Xy(X)
         if X.shape[-1] != len(self.estimators_):
             raise ValueError('The number of estimators does not match '
-                             'X.shape[2]')
+                             'X.shape[-1]')
         # For predictions/transforms the parallelization is across the data and
         # not across the estimators to avoid memory load.
         parallel, p_func, n_jobs = parallel_func(_sl_score, self.n_jobs)
+        n_jobs = np.min([n_jobs, X.shape[-1]])
```
Comment (agramfort) [inline mne/decoding/search_light.py]: idem

File: mne/decoding/search_light.py [hunk 5]
```diff
@@ -371,9 +392,10 @@ def _transform(self, X, method):
         self._check_Xy(X)
         method = _check_method(self.base_estimator, method)
         parallel, p_func, n_jobs = parallel_func(_gl_transform, self.n_jobs)
+        n_jobs = np.min([n_jobs, X.shape[-1]])
```
Comment (agramfort) [inline mne/decoding/search_light.py]: idem

File: mne/decoding/search_light.py [hunk 6]
```diff
@@ -453,32 +483,29 @@ def decision_function(self, X):
...
-
-        Notes
-        -----
-        This requires base_estimator to have a `decision_function` method.
         """
         self._check_Xy(X)
         # For predictions/transforms the parallelization is across the data and
         # not across the estimators to avoid memory load.
         parallel, p_func, n_jobs = parallel_func(_gl_score, self.n_jobs)
+        n_jobs = np.min([n_jobs, X.shape[-1]])
```
Comment (agramfort) [inline mne/decoding/search_light.py]: idem

File: mne/decoding/tests/test_search_light.py [hunk 7]
```diff
@@ -71,9 +76,20 @@ def transform(self, X):
     pipe.fit(X, y)
     pipe.predict(X)
 
+    # n-dimensional feature space
+    X = np.random.rand(10, 3, 4, 2)
+    y = np.arange(10) % 2
+    for n_jobs in [1, 2]:
+        pipe = SearchLight(make_pipeline(Vectorizer(), LogisticRegression()),
+                           n_jobs=n_jobs)
+        pipe.fit(X, y).predict(X)
```
Comment (agramfort) [inline mne/decoding/tests/test_search_light.py]: can you test at least the same of the prediction?

File: mne/decoding/tests/test_search_light.py [hunk 8]
```diff
@@ -108,3 +124,17 @@ def test_generalizationlight():
...
+    gl.fit(X[..., [0]], y)
+    gl.predict(X[..., [0]])
+
+    # n-dimensional feature space
+    X = np.random.rand(10, 3, 4, 2)
+    y = np.arange(10) % 2
+    for n_jobs in [1, 2]:
+        pipe = GeneralizationLight(
+            make_pipeline(Vectorizer(), LogisticRegression()), n_jobs=n_jobs)
+        pipe.fit(X, y).predict(X)
```
Comment (agramfort) [inline mne/decoding/tests/test_search_light.py]: can you test at least the same of the prediction?

**Comment by agramfort** [comment]
besides LGTM

**Comment by agramfort** [comment]
thx @kingjr

**Comment by kingjr** [comment]
thanks for the reviews!

**Pull request merged**